### PR TITLE
Fix standard deviation gradient

### LINKED ISSUE #4320: torch.std NaN gradient
`torch.std(x)` gives NaN gradients at `x=0`.

```py
import torch
from torch.autograd.variable import Variable
import torch.nn.functional as F

a = Variable(torch.FloatTensor(16,3,64,64).zero_(),requires_grad=True)

# b = a.std(dim=0).mean()  # NaN gradients
# b = (a - a.mean(dim=0,keepdim=True)).norm(p=2,dim=0).mean() / (15**0.5)  # 0 gradients

b.backward()

print(a.grad.data)
```

related #2421 #3264

cc @ezyang @SsnL @albanD @zou3519 @gqchen @gchanan

### PR DESCRIPTION
**Pull request by vishwakftw**
Defined the subgradient of `std` when `result = 0` to be `inf`.

1. Added tests in test_autograd
2. New method std_backward() for computing backward in Functions.cpp

Fixes #4320 .

### CODE REVIEW

**Comment by SsnL** [comment]
:( this adds 2 extra ops

**Comment by vishwakftw** [comment]
I can make it one extra op alone, by modifying the entry in `Declarations.cwrap` alone.

File: tools/autograd/derivatives.yaml [hunk 1]
```diff
@@ -648,10 +648,10 @@
   self: unsqueeze_to(grad, dim, self.sizes())
 
 - name: std(Tensor self, bool unbiased)
-  self: std_backward(grad, self, result, unbiased)
+  self: var_backward(grad / (2 * result), self, unbiased).masked_fill_(result == 0., INFINITY)
```
Comment (vishwakftw) [inline tools/autograd/derivatives.yaml]: I think I understand your concern: `masked_fill` is one op, and `result == 0.` is another op.

Comment (vadimkantorov) [reply tools/autograd/derivatives.yaml]: Another instance where `torch.where` special case for float zeros may be useful? (although it's hacky) https://github.com/pytorch/pytorch/pull/8945#issuecomment-401211309

Comment (ezyang) [reply tools/autograd/derivatives.yaml]: Yeah, this is really one of those situations where we should just JIT the kernel and everything would fuse just fine...

**Comment by ezyang** [comment]
@vishwakftw I'm perhaps not the best core dev to ask about this, since I have always leaned on the side of correctness versus perf, but maybe it would help appease fears about slowness to do a quick benchmark before and after to see what the perf impact is.

**Comment by vishwakftw** [comment]
This is the time difference:
```
# Without masked_fill_(result == 0.0, INFINITY)
81 ms ± 1.71 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

# With masked_fill_(result == 0.0, INFINITY)
128 ms ± 2.65 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

# With recent version (optimized masked_fill)
109 ms ± 1.87 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

Regarding use of `inf` instead of of any other value in `-inf, inf`, I did this to match the behaviour of `sqrt`.

**Review by colesbury — CHANGES_REQUESTED** [review]
`inf` is not a good choice for the gradient of std() when std() is 0.

You can derive a formula using the limit definition of a derivative and the formula for standard deviation. It's going to depend on the number of elements `N`.

**Comment by colesbury** [comment]
I think the formula should be:

unbiased: `1/sqrt(n)`
biased: `sqrt(n-1)/n`

You can verify this numerically

```
N = 7
eps = 1e-9
x = torch.zeros(N, dtype=torch.double)
x[0] = eps
print(float(torch.std(x) / eps))
print(float(1/math.sqrt(N)))
```

**Comment by vishwakftw** [comment]
@colesbury Thanks for the advice. I computed the derivative using the definition and it came out to be:
`sqrt(n-1)/nN` where `N = n-1` for unbiased and `n` otherwise.

**Comment by SsnL** [comment]
@colesbury then you can apply the same logic and say that they should be
negated ones of what you have.

**Comment by SsnL** [comment]
I think it depends on which direction you take the limit.

**Comment by SsnL** [comment]
Hmmm isn’t stddev symmetrical at origin?

**Comment by vishwakftw** [comment]
If we take `x - delta`, then derivative is `sqrt((1 + 3/n)/N)` where `N` is `n-1` if unbiased, `n` if biased.

File: tools/autograd/templates/Functions.cpp [hunk 2]
```diff
@@ -575,6 +575,28 @@ Tensor var_backward(Tensor grad, const Tensor & self, int64_t dim, bool unbiased
   return (2.0 / (self.size(dim) - unbiased)) * grad * (self - self.mean(dim, true));
 }
 
+Tensor std_backward(const Tensor & grad, const Tensor & self, Tensor result, bool unbiased) {
+  Tensor result_zero_mask = (result == 0.);
+  if (result_zero_mask.any().toCByte()) {
```
Comment (ezyang) [inline tools/autograd/templates/Functions.cpp]: Hmm... in the CUDA case, I wonder if it wouldn't be better to unconditionally apply the masked fill, because that lets you prevent a synchronization.

Comment (ezyang) [reply tools/autograd/templates/Functions.cpp]: (You didn't specify if your perf experiment was CPU or CUDA)

**Comment by ezyang** [comment]
@SsnL is the math good now?

File: tools/autograd/templates/Functions.cpp (hunk 2 above)
Comment (vishwakftw) [reply tools/autograd/templates/Functions.cpp]: The perf experiment was CPU. Sorry about missing this out.

**Comment by SsnL** [comment]
Derivative is not the same if we take the limit from `x->0-` direction. The change in the output is the same, but the change in the input is negated. So the limit is negative of that taken from `x->0+` direction. You can verify this using @colesbury 's code above but with `eps=-1e-9`. So essentially anywhere in range `[-1/sqrt(N), 1/sqrt(N)]` is valid subgradient.

Since the usage is mostly gradient-based optimization like gd, it makes sense to take one of these two extremes. But it's unclear to me that which we should choose. (`nan` really isn't a bad choice in this sense.) I think if we choose one, we should state the choice in the doc.

**Comment by vishwakftw** [comment]
I thought we were waiting for @colesbury ‘s call. If not, I am sorry, I will send in the changes as soon as I can.

**Comment by ezyang** [comment]
No, I'm sorry; I wasn't sure what we were waiting on. If you like I can bug him about it tomorrow.

**Comment by vishwakftw** [comment]
@ezyang Here are the timings you were curious about:
```
With CUDA
With unconditional masked_fill_ in std_backward() for tensor of size 1000 x 1000 x 10 filled with 1s
33.6 ms ± 173 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

With conditional masked_fill_ in std_backward() for tensor of size 1000 x 1000 x 10 filled with 1s
33.7 ms ± 275 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

With unconditional masked_fill_ in std_backward() for tensor of size 1000 x 1000 x 10 filled using randn
31.2 ms ± 111 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

With conditional masked_fill_ in std_backward() for tensor of size 1000 x 1000 x 10 filled using randn
30.8 ms ± 209 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

`conditional masked_fill_` is the current implementation.

**Comment by vishwakftw** [comment]
Is this good to go?

**Comment by colesbury** [comment]
I'm a bit wary about this change. It looks like it introduces a CUDA synchronization point in result_zero_mask.any().toCByte() (as ezyang pointed out).

synchronization points may not slow down an individual call, but they make it harder to hide kernel launch latency and CPU computation in a bigger system.

**Comment by vishwakftw** [comment]
Are you suggesting that the `masked_fill_` be done unconditionally?

**Comment by ezyang** [comment]
I hate to say it, but maybe we should just write the fused kernel that does the masked fill unconditionally. Then we don't have to worry about the extra kernel launch overhead.

**Comment by vishwakftw** [comment]
This is stale, closing for now.
